### ISSUE TITLE
Parachain runtime upgrade is triggering stall due to peer reputation issues

### ISSUE THREAD

**Issue by crystalin** [paritytech/polkadot#4617]
This is happening in multiple networks (including public Moonbase Alphanet, but the impact is lower there). I'll focus on our internal Moonsilver network composed of 4 validators (v0.9.13, runtime 9140, 2 max validators per core) and 8 collators (in 2 different world regions).

When performing a Runtime upgrade, the parachain stops producing blocks for 5-10 minutes, and then slowly 2 collators started to produce blocks. the 6 other collators never produced block again (for at least 1 hour) until we restarted them.

Here is my hypothesis :
- Runtime upgrade is creating a special condition (higher than usual PoV, + maybe some extra work on the collator/validator.
- That block is not able to be fetched in time by the validators. After many, many tries, one of the collator succeeds
- Validators are reducing peer reputation of the collator because of "Timeout"
- Collators end up being "banned" but are still trying to produce blocks
- "Banned" collators are trying to connect again after producing a block, which gets refused by the validator, which reduce even more their reputation
- This snowballs and the collators are never able to connect again and produce blocks

Graph of the connections:
![image (2)](https://user-images.githubusercontent.com/329248/147518715-dad7277f-3be4-4e89-8619-8373e87a92d1.png)

Logs of the 8 collators + 4 validators (1h) at the time of the upgrade:
https://drive.google.com/file/d/1d6ig76i1Cg6NAKcAKI4e5tqxpC7WwhvE/view?usp=sharing

(This same scenario happens on Moonbase Alphanet, however the stall is usually less long (due to diversity of collator locations) and less collators are impacted (because the stall is less long I suppose))

**Comment by crystalin** [comment]
@eskimor (Basti told me to mention you :) )

**Comment by crystalin** [comment]
Also, to add to the hypothesis, I think the impact is less on alphanet, because having 80 collators, the time to get selected to produce the next block (and so get punished again if you are already "banned") is longer and so allows to recover reputation faster than you lose it.... maybe ?

**Comment by eskimor** [comment]
Thanks @crystalin - we will look into it!

**Comment by crystalin** [comment]
On our Alphanet network, (that was impacted by this same issue), we deployed a version of the relay that is not punishing for collation fetching issues: https://github.com/PureStake/polkadot/commit/dfd86c00a260fa4dec609d85e8e7db2d821ce5ae

The collators were all able to recover few hours after as you can see from this block production time chart:
![image](https://user-images.githubusercontent.com/329248/147619182-77b30408-cdff-4fb8-88a4-45efba308233.png)

I'm not sure it is what fixed it but could be

**Comment by eskimor** [comment]
You also increased the timeout, which would also help in getting the block in eventually. If you restart the validator nodes, peer reputations would be reset  - so I don't really see how the peer reputation change alone would explain this.

The thing is, if you are hitting the timeout, you are already pushing the boundaries of the system a lot. If the block was big because of a runtime upgrade, then the actual backing is actually really half the work, if at all: After the backing group validated the candidate, they distribute their statements and those statements, which need to reach the block producer, will contain that runtime upgrade! If the runtime upgrade is megabytes in size, and you fully exhausted the new timeout of 1.5 seconds, assuming it only takes 0.1 seconds to validate the candidate and to produce it, then we would have a total of 0.3 seconds to distribute megabytes of data to a thousand validators (on Kusama). 

I think the reason why this will work eventually is, because we don't really need to reach all validators, we only need to reach the block producer, so by chance the block producer will be close or maybe one of the backing validators themselves, which will be the time when the block inclusion finally succeeds. 

That's why a larger timeout does not really solve the problem, although it should help in getting the candidate in at least at some point.

Good news is, the PR of reducing the number of required votes is already on its way and should help with those issues (it also increases the timeout a bit).

**Comment by crystalin** [comment]
@eskimor yes I increased the timeout as it helped to pass the PoV without getting punished for the timeout (The collation takes multiple tcp rounds before sending the PoV, making 200ms+ latency impossible to send the PoV in less than 1000ms in default tcp configuration for most linux dist). 

However the issue here is not the fact the runtime upgrade (or any other big block) doesn't get in. It eventually does (and you can see in the logs or in the chart that it is the same list of collators that were producing the blocks, until we did the hotfix on the relay), but then all the collators that were not able to include that block are "banned" **forever** (at least in small networks). The reason is because they keep getting punished while trying to send new blocks, because if you are "banned" and you try to connect, you get punished more or something similar

**Comment by eskimor** [comment]
With small network you also mean a small relay chain network - correct? Because on a large network a permanent ban should not happen that easily, because validator groups rotate, so after 2 minutes you will connect to a new validator set that has not heard from you in a while and should thus not have you banned at all. That means in a network where there is at least a couple of validator groups, a ban should not be longer than two minutes/rotation frequency.

In any case an update is on its way, which increases the timeout a bit and reduces the number of required backing votes. This means, if validators on slower connections ban the collator it should matter less: If it is able to deliver its collation at least to two validators out of five - up to 3 slow validators can ban a collator, without causing issues.

We will also be looking into getting rid of banning for timeouts completely, as an interim solution.

**Comment by eskimor** [comment]
> When performing a Runtime upgrade, the parachain stops producing blocks for 5-10 minutes, and then slowly 2 collators started to produce blocks. the 6 other collators never produced block again (for at least 1 hour) until we restarted them.

I am not sure how a restart of the collators would fix an issue with reputations. Do the collators get a new `PeerId` when they are restarted or have they been shut down for a longer period of time?

**Comment by crystalin** [comment]
@eskimor yes, I'm also surprised by the restart, and it is not always consistent. What we usually do is to restart the collators and if it doesn't work we restart the validators. It usually end up fixing it.
